EIPS/eip-3540.md

```diff
@@ -301,45 +301,57 @@ S_DATA = 0x02
 def is_eof(code: bytes) -> bool:
     return code.startswith(MAGIC)
 
+class ValidationException(Exception):
+    pass
 
-# Validate EOF code.
+# Raises ValidationException on invalid code
 def validate_eof(code: bytes):
     # Check version
-    assert len(code) >= 3 and code[2] == VERSION
+    if len(code) < 3 or code[2] != VERSION:
+        raise ValidationException("invalid version")
 
     # Process section headers
     section_sizes = {S_CODE: 0, S_DATA: 0}
     pos = 3
     while True:
         # Terminator not found
-        assert pos < len(code)
+        if pos >= len(code):
+            raise ValidationException("no section terminator")            
+        
         section_id = code[pos]
         pos += 1
         if section_id == S_TERMINATOR:
             break
 
         # Disallow unknown sections
-        assert section_id in section_sizes
+        if not section_id in section_sizes:
```

Comment:
use assert with message
```suggestion
        assert (section_id in section_sizes), "invalid section id"
```

Reply:
This would work, but two (arguably minor) reasons to keep it with `raise`:

1. Being more explicit that we raise an exception (i.e. don't rely on reader's knowledge that `assert` in python raises `AssertionError`)
2. [execution-specs](https://github.com/ethereum/execution-specs) uses [hierarchy of exceptions](https://github.com/ethereum/execution-specs/blob/master/src/ethereum/berlin/vm/exceptions.py) to report errors, and we envision this code might be integrated there eventually, so we see a benefit in keeping the style close.

Reply:
(note also that we keep this code in https://github.com/ipsilon/eof/tree/main/eips_code and have extensive test suite for it there)